File: rust/realtime/src/filters.rs
//! Filters

use ndarray::prelude::*;
use ndarray::stack;



/// Dummy 'filter' that returns unmodified input as 'prediction'
pub struct FilterAutoEcho {}

/// Trait for autoregressive adaptive filters
pub trait FilterAuto {
    fn predict_next(&mut self, x: &Array1<f64>) -> Array1<f64>;
}

impl FilterAuto for FilterAutoEcho {

    fn predict_next(&mut self, x: &Array1<f64>) -> Array1<f64> {
        return x.clone()
    }
}


/// Least-mean-squares AdaptiveFilter (needed for FilterAutoLMS)
pub struct FilterLMS {
    mu: f64,
    wts: Array2<f64>,
}

impl FilterLMS {

    pub fn new(dim: u16, order: u16, mu:f64) -> FilterLMS {

        let wts_shape = (dim as usize, (order * dim) as usize);

        FilterLMS {
            mu: mu,
            wts: Array::zeros(wts_shape),
        }
    }

    /// Adapt filter to new (desired, input) pair
    pub fn adapt(&mut self, d: &Array2<f64>, x: &Array2<f64>) {

        // Flatten input matrix into vector
        let x_vec = x.t().into_shape((self.wts.ncols(), 1)).unwrap();

        // Compute weight update
        let y: Array2<f64> = self.wts.dot(&x_vec);
        let e: Array2<f64> = d - &y;
        let wts_update = self.mu * e.dot(&x_vec.t());

        self.wts += &wts_update;
    }

    /// Predict value for input
    pub fn predict(&self, x: &Array2<f64>) -> Array2<f64> {

        let x_vec = x.t().into_shape((self.wts.ncols(), 1)).unwrap();
        self.wts.dot(&x_vec)
    }
}
    

/// FilterAuto that uses LMS to filter autoregressive process
pub struct FilterAutoLMS {
    flt_lms: FilterLMS,
    x_hist: Array2<f64>,
}

impl FilterAutoLMS {

    pub fn new(dim: u16, order: u16, mu: f64) -> FilterAutoLMS {
        FilterAutoLMS{
            flt_lms: FilterLMS::new(dim, order, mu),
            x_hist: Array::zeros((dim as usize, order as usize)),
        }
    }
}

impl FilterAuto for FilterAutoLMS {

    fn predict_next(&mut self, x: &Array1<f64>) -> Array1<f64> {

        // Convert x to 2D array
        let x_2d = x.clone().insert_axis(Axis(1));
        
        // Update filter using history as input and current value as output
        self.flt_lms.adapt(&x_2d, &self.x_hist);

        // Add current value to history, dropping oldest value 
        let x_hist_trunc = self.x_hist.slice(s![.., 0..-1]);
        self.x_hist = stack![Axis(1), x_2d, x_hist_trunc];

        // Use new history to predict next value
        let y_2d = self.flt_lms.predict(&self.x_hist);
        let y = y_2d.index_axis_move(Axis(1), 0);

        return y;
    }
}

File: c/realtime/src/filters.c
/* Filters */

#include <stdlib.h>
#include <stdio.h>
#include <cblas.h>

#include "filters.h"


// Constructor for FilterAutoLMS object
void FilterAutoLMS_new(struct FilterAutoLMS* flt, int dim, int order, double mu) {

    // Number of elements in history
    int hist_size = dim * order;

    // Allocate array for prediction and set to zero
    flt->x_pred = (double *) malloc(dim * sizeof(double));
    for (int i = 0; i < dim; i++) {
        flt->x_pred[i] = 0.0;
    }

    // Allocate array for scaled error and set to zero
    flt->x_err = (double *) malloc(dim * sizeof(double));
    for (int i = 0; i < dim; i++) {
        flt->x_err[i] = 0.0;
    }

    // Allocate array for history and set to zero 
    flt->x_hist = (double *) malloc(hist_size * sizeof(double));
    for (int i = 0; i < hist_size; i++) {
        flt->x_hist[i] = 0.0;
    }

    // Allocate array for weights and set to zero
    flt->wts = (double *) malloc(dim * hist_size * sizeof(double));
    for (int i = 0; i < dim * hist_size; i++) {
        flt->wts[i] = 0.0;
    }

    // Populate fields
    flt->dim = dim;
    flt->order = order;
    flt->hist_size = hist_size;
    flt->mu = mu;
}


// Destructor for FilterAutoLMS object
void FilterAutoLMS_delete(struct FilterAutoLMS* flt) {

    // Free allocated memory
    free(flt->wts);
    free(flt->x_hist);
    free(flt->x_err);
    free(flt->x_pred);
}


// Update filter with new signal value and predict next value (using CBLAS)
void FilterAutoLMS_predict_next(struct FilterAutoLMS* flt, double* x) {

    // Compute error
    for (int i = 0; i < flt->dim; i++) {
        flt->x_err[i] = x[i] - flt->x_pred[i];
    }

    // Update weights (wts := wts + mu * x_err * x_hist')
    cblas_dger(
        CblasRowMajor, flt->dim, flt->hist_size, flt->mu, 
        flt->x_err, 1, flt->x_hist, 1, flt->wts, flt->hist_size
    );

    // Update history (x_hist[dim:end] := x_hist[0:end-dim]; x_hist[0:dim] := x)
    for (int i = (flt->hist_size - 1); i >= flt->dim; i--) {
        flt->x_hist[i] = flt->x_hist[i - flt->dim];
    }
    for (int i = 0; i < flt->dim; i++) {
        flt->x_hist[i] = x[i];
    }

    // Update prediction (x_pred := wts * x_hist)
    cblas_dgemv(
        CblasRowMajor, CblasNoTrans, flt->dim, flt->hist_size, 1.0, 
        flt->wts, flt->hist_size, flt->x_hist, 1, 0.0, flt->x_pred, 1
    );
}


// Constructor for FilterAutoEcho object
void FilterAutoEcho_new(struct FilterAutoEcho* flt, int dim) {

    // Allocate array for prediction and set to zero
    flt->x_pred = (double *) malloc(dim * sizeof(double));
    for (int i = 0; i < dim; i++) {
        flt->x_pred[i] = 0.0;
    }

    // Populate fields
    flt->dim = dim;
}


// Destructor for FilterAutoEcho object
void FilterAutoEcho_delete(struct FilterAutoEcho* flt) {

    // Free allocated memory
    free(flt->x_pred);
}


// Update filter with new signal value and predict next value
void FilterAutoEcho_predict_next(struct FilterAutoEcho* flt, double* x) {

    // Copy input to prediction
    for (int i = 0; i < flt->dim; i++) {
        flt->x_pred[i] = x[i];
    }
}
